### PR TITLE
Use wayland::Weak in XDG shell implementations

### PR DESCRIPTION
**Pull request by wmww**
Replace the complicated and error prone direct usage of destroyed flags with `wayland::Weak`

### CODE REVIEW

File: src/server/frontend_wayland/xdg_shell_stable.cpp [hunk 1]
```diff
@@ -58,16 +58,15 @@ class XdgSurfaceStable : wayland::XdgSurface
 
     void send_configure();
 
-    std::experimental::optional<WindowWlSurfaceRole*> const& window_role();
+    wayland::Weak<WindowWlSurfaceRole> const& window_role();
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/xdg_shell_stable.cpp:61]: The `wayland::` prefix looks redundant.

File: src/server/frontend_wayland/xdg_shell_stable.cpp [hunk 2]
```diff
@@ -258,24 +261,19 @@ void mf::XdgSurfaceStable::send_configure()
...
-
     return window_role_;
 }
 
 void mf::XdgSurfaceStable::set_window_role(WindowWlSurfaceRole* role)
 {
-    if (window_role())
+    if (window_role_)
+    {
         log_warning("XdgSurfaceStable::window_role set multiple times");
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/xdg_shell_stable.cpp:273]: Pre-existing, but we ought to provide some information about which client & window is misbehaving.

Comment (wmww) [reply src/server/frontend_wayland/xdg_shell_stable.cpp:273]: The correct behavior here is to send a Wayland error and kill the client, but knowing what error to send will take a closer look. I believe there may already be a WLCS test for this (at least there are tests for similar things). As you said, pre-existing. I'll look at that another day.

**Pull request merged**